### ISSUE TITLE
MIXED PRECISION TRAINING

### ISSUE THREAD

**Issue by joapolarbear** [joapolarbear/dl_notes#17]
### ICLR 2018
[PDF](https://arxiv.org/pdf/1710.03740.pdf)

# Main Idea
address the **training** with reduced precision while **maintaining model accuracy**

# Three techniques to prevent model accuracy loss

## FP32 mater copy of weights
maintaining a single-precision copy of weights that accumulates the gradients after each optimizer step (this copy is rounded to half-precision for the forward- and back-propagation).

![image](https://user-images.githubusercontent.com/17765864/87009057-3e07c180-c1f7-11ea-97fe-82c7a3d5fbd9.png)
**Weight updates (weight gradients * learning rate)** are stored in FP32**, so there is an **assumption** that the gradients are > 2^{-14}, but the _weight updates_ tend to be smaller than 2^{-14}, thus become 0.

Another reason, if  the ratio of the weight value to the weight update is very large, the update would become zero

## Loss Scaling
propose loss-scaling to preserve gradient values with small magnitudes. 

## Arithmetic Precision
* Apart from activations, weights, gradients, here discuss reducing precision of _arithmetic_.
* use half-precision arithmetic that accumulates into **single-precision** outputs, which are converted to **half-precision** before storing to memory.

Three arithmetic categories: 
### 1. vector dot-products
sensitive to arithmetic speed 👉🏻 use fp16
### 2. reduction
> Large reductions (sums across elements of a vector) should be carried out in FP32 
- [ ] > This did not slow down the training process since these layers are memory-bandwidth limited and not sensitive to arithmetic speed.

### 3. point-wise operation
memory- bandwidth limited, either FP16 or FP32 math can be used.
